### PR TITLE
#25849 simplify database generation (v2)

### PR DESCRIPTION
**Pull request by danchr**
To ease review, this is an incremental change to #94, the database generation PR. The change gets rid of all the generated files, and replaces them with a single test fixture to ease future changes to the templates. The advantage of a single file over multiple is that it is both easier to inspect if needed — and ignore, if not.

The change also redirects all uses of `db_structure` to go through settings, so that we may override it during deployment, for example.

### CODE REVIEW

File: doc/database_changes.rst [hunk 1]
```diff
@@ -1,6 +1,10 @@
 Working with database templates and patches for Magenta's OIO/REST framework
 ============================================================================
 
+.. ATTENTION::
+   This documentation is outdated; We no longer rely on patches and
```
Comment (agger-magenta) [inline doc/database_changes.rst]: Well, why not update it rather than putting up a note that it's outdated?

Comment (danchr) [reply doc/database_changes.rst]: Well, the thing is… With @carlbordum's recent changes, we no longer use the patches that this document describes. We should probably just delete this file, but I figured at least improved the state of matters somewhat 😕

Comment (agger-magenta) [reply doc/database_changes.rst]: You're right. I think the right thing to do is delete this file. We still have some way to go in order to make the work with model changes painless, but this patchwork is no longer a part of our system, so let's relegate it to (git) History.

File: oio_rest/oio_rest/db_layout.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+"""
+    apply-templates.py
```
Comment (carlbordum) [inline oio_rest/oio_rest/db_layout.py]: new name.

What happened to all the  stuff you added in the preceding pr? The `click` interface?

File: oio_rest/oio_rest/db_layout.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        for template_name in TEMPLATES:
+            template_file = "%s.jinja.sql" % template_name
+            template = template_env.get_template(template_file)
+
+            context = copy.deepcopy(db_structure[oio_type])
+            context["script_signature"] = "apply-template.py %s %s" % (
+                oio_type,
+                template_file,
+            )
+            # it is important that the order is stable, as some templates rely on this
```
Comment (carlbordum) [inline oio_rest/oio_rest/db_layout.py]: I think this comment may be lying.

File: db/db-templating/templates/as_list.jinja.sql [hunk 4]
```diff
@@ -199,7 +199,7 @@ FROM
 					FROM		{{oio_type}} a
 					JOIN 		{{oio_type}}_registrering b 	ON b.{{oio_type}}_id=a.id
 					WHERE a.id = ANY ({{oio_type}}_uuids) AND ((registrering_tstzrange is null AND upper((b.registrering).timeperiod)='infinity'::TIMESTAMPTZ) OR registrering_tstzrange && (b.registrering).timeperiod)--filter ON registrering_tstzrange
-				{%-for attribut , attribut_fields in attributter_revorder.items() %}{% set outer_loop = loop %}
+				{%-for attribut , attribut_fields in attributter.items() | reverse %}{% set outer_loop = loop %}
```
Comment (carlbordum) [inline db/db-templating/templates/as_list.jinja.sql]: Thanks you for these :)

File: oio_rest/oio_rest/db_layout.py (hunk 2 above)
Comment (agger-magenta) [inline oio_rest/oio_rest/db_layout.py]: Hvorfor skriver du i kommentaren, at den hedder `apply-templates`, når den faktisk hedder `db_layout`?

Og vil `apply_templates` (hvis den skal bruges som modul) måske stadig være et mere retvisende navn?

Comment (danchr) [reply oio_rest/oio_rest/db_layout.py]: The `click` interface became redundant once we no longer had any `generated-files` :)

File: oio_rest/oio_rest/db_layout.py (hunk 3 above)
Comment (danchr) [reply oio_rest/oio_rest/db_layout.py]: > I think this comment may be lying.

I'm afraid not 😞 With the following change:

```diff
diff --git a/oio_rest/oio_rest/db_layout.py b/oio_rest/oio_rest/db_layout.py
--- a/oio_rest/oio_rest/db_layout.py
+++ b/oio_rest/oio_rest/db_layout.py
@@ -52,9 +52,8 @@ def render_templates():
                 oio_type,
                 template_file,
             )
-            # it is important that the order is stable, as some templates rely on this
-            context["tilstande"] = OrderedDict(context["tilstande"])
-            context["attributter"] = OrderedDict(context["attributter"])
+            context["tilstande"] = dict(context["tilstande"])
+            context["attributter"] = dict(context["attributter"])
             context["oio_type"] = oio_type.lower()
 
             try:
```

…and the following command:

```
PYTHONHASHSEED=0 python3.5 -m unittest -v tests.test_sql tests.test_rest_indsats
```

I get this:

```test_pgsql (tests.test_sql.SQLTests) ... ERROR
test_sql_unchanged (tests.test_sql.SQLTests) ... FAIL
test_indsats_create (tests.test_rest_indsats.TestIndsats) ... …/oio_rest/tests/test_rest_indsats.py:13: ResourceWarning: unclosed file <_io.TextIOWrapper name='tests/fixtures/indsats_opret.json' mode='rt' encoding='UTF-8'>
  "json": open("tests/fixtures/indsats_opret.json", "rt").read(),
ERROR
test_indsats_put (tests.test_rest_indsats.TestIndsats) ... …/oio_rest/tests/test_rest_indsats.py:25: ResourceWarning: unclosed file <_io.TextIOWrapper name='tests/fixtures/indsats_opret.json' mode='rt' encoding='UTF-8'>
  "json": open("tests/fixtures/indsats_opret.json", "rt").read(),
ERROR

======================================================================
ERROR: test_pgsql (tests.test_sql.SQLTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "…/oio_rest/tests/test_sql.py", line 43, in test_pgsql
    curs.execute("SELECT * FROM runtests ('test'::name)")
psycopg2.ProgrammingError: cannot cast type record to indsatsregistreringtype
LINE 12: ...dsatsRelIndsatsaktoer1,indsatsRelIndsatsaktoer2]) :: indsats...
                                                              ^
DETAIL:  Cannot cast type indsatspublicerettilstype[] to indsatsfremdrifttilstype[] in column 2.
QUERY:  SELECT ROW (

	ROW (
	NULL,
	'Opstaaet'::Livscykluskode,
	uuidRegistrering,
	'Test Note 4') :: RegistreringBase
	,
	ARRAY[indsatsPubliceret]::IndsatsPubliceretTilsType[],
ARRAY[indsatsFremdrift]::indsatsFremdriftTilsType[],
ARRAY[indsatsEgenskab]::indsatsEgenskaberAttrType[],
ARRAY[indsatsRelIndsatsmodtager,indsatsRelIndsatssag1,indsatsRelIndsatssag2,indsatsRelIndsatsaktoer1,indsatsRelIndsatsaktoer2]) :: indsatsRegistreringType
CONTEXT:  PL/pgSQL function test.test_as_create_or_import_indsats() line 189 at assignment
PL/pgSQL function _runner(text[],text[],text[],text[],text[]) line 62 at FOR over EXECUTE statement
SQL function "runtests" statement 1
SQL function "runtests" statement 1


======================================================================
ERROR: test_indsats_create (tests.test_rest_indsats.TestIndsats)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "…/oio_rest/tests/test_rest_indsats.py", line 13, in test_indsats_create
    "json": open("tests/fixtures/indsats_opret.json", "rt").read(),
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 840, in post
    return self.open(*args, **kw)
  File "…/python-testenv/lib/python3.5/site-packages/flask/testing.py", line 200, in open
    follow_redirects=follow_redirects
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 803, in open
    response = self.run_wsgi_app(environ, buffered=buffered)
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 716, in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 923, in run_wsgi_app
    app_rv = app(environ, start_response)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 2295, in wsgi_app
    response = self.handle_exception(e)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1741, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "…/python-testenv/lib/python3.5/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "…/python-testenv/lib/python3.5/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "…/oio_rest/oio_rest/authentication.py", line 104, in decorated
    return f(*args, **kwargs)
  File "…/oio_rest/oio_rest/oio_rest.py", line 204, in create_object
    uuid = db.create_or_import_object(cls.__name__, note, registration)
  File "…/oio_rest/oio_rest/db.py", line 427, in create_or_import_object
    cursor.execute(sql)
psycopg2.ProgrammingError: cannot cast type record to indsatsregistreringtype
LINE 77:         ) :: IndsatsRegistreringType,
                   ^
DETAIL:  Cannot cast type indsatspublicerettilstype[] to indsatsfremdrifttilstype[] in column 2.


======================================================================
ERROR: test_indsats_put (tests.test_rest_indsats.TestIndsats)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "…/oio_rest/tests/test_rest_indsats.py", line 25, in test_indsats_put
    "json": open("tests/fixtures/indsats_opret.json", "rt").read(),
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 850, in put
    return self.open(*args, **kw)
  File "…/python-testenv/lib/python3.5/site-packages/flask/testing.py", line 200, in open
    follow_redirects=follow_redirects
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 803, in open
    response = self.run_wsgi_app(environ, buffered=buffered)
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 716, in run_wsgi_app
    rv = run_wsgi_app(self.application, environ, buffered=buffered)
  File "…/python-testenv/lib/python3.5/site-packages/werkzeug/test.py", line 923, in run_wsgi_app
    app_rv = app(environ, start_response)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 2309, in __call__
    return self.wsgi_app(environ, start_response)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 2295, in wsgi_app
    response = self.handle_exception(e)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1741, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "…/python-testenv/lib/python3.5/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 2292, in wsgi_app
    response = self.full_dispatch_request()
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1815, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1718, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "…/python-testenv/lib/python3.5/site-packages/flask/_compat.py", line 35, in reraise
    raise value
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1813, in full_dispatch_request
    rv = self.dispatch_request()
  File "…/python-testenv/lib/python3.5/site-packages/flask/app.py", line 1799, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "…/oio_rest/oio_rest/authentication.py", line 104, in decorated
    return f(*args, **kwargs)
  File "…/oio_rest/oio_rest/oio_rest.py", line 366, in put_object
    db.create_or_import_object(cls.__name__, note, registration, uuid)
  File "…/oio_rest/oio_rest/db.py", line 427, in create_or_import_object
    cursor.execute(sql)
psycopg2.ProgrammingError: cannot cast type record to indsatsregistreringtype
LINE 77:         ) :: IndsatsRegistreringType,
                   ^
DETAIL:  Cannot cast type indsatspublicerettilstype[] to indsatsfremdrifttilstype[] in column 2.


======================================================================
FAIL: test_sql_unchanged (tests.test_sql.SQLTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "…/oio_rest/tests/test_sql.py", line 75, in test_sql_unchanged
    'contents changed -- new dump is in {}'.format(actual_path),
AssertionError: '-- C[502841 chars]dsatsPubliceretTils AS ENUM (\'Publiceret\',\'[1169625 chars]\n\n' != '-- C[502841 chars]dsatsFremdriftTils AS ENUM (\'Uoplyst\',\'Visi[1169627 chars]\n\n' : contents changed -- new dump is in …/oio_rest/tests/fixtures/db-dump.sql.new

----------------------------------------------------------------------
Ran 4 tests in 8.077s

FAILED (failures=1, errors=3)
```

Comment (carlbordum) [reply oio_rest/oio_rest/db_layout.py]: Ew. We'll get there...

File: oio_rest/oio_rest/db_templating.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    db_structure = settings.DB_STRUCTURE.DATABASE_STRUCTURE
+    extra_options = settings.DB_STRUCTURE.DB_TEMPLATE_EXTRA_OPTIONS
+
+    for oio_type in sorted(db_structure):
+        for template_name in TEMPLATES:
+            template_file = "%s.jinja.sql" % template_name
+            template = template_env.get_template(template_file)
+
+            context = copy.deepcopy(db_structure[oio_type])
+            context["script_signature"] = "apply-template.py %s %s" % (
```
Comment (agger-magenta) [inline oio_rest/oio_rest/db_templating.py]: Hæh, busted!

**Pull request merged**